docs/index.md

```diff
@@ -0,0 +1,268 @@
+   do this with anything else in our codebase!
+1. It puts the error-handling right up front and not in a good way. While we want to have
+   a failure case in mind when designing the behavior of our functions, it’s not usually
+   the point of most functions – things like `HandleErr` in the above example being the
+   exception and not the rule. The actual meat of the function is always after the error
+   handling.
+
+But if we’re not using some similar kind of callback pattern, we usually resort to
+exceptions. But exceptions are unpredictable: you can’t know whether a given function
+invocation is going to throw an exception until runtime as someone calling the function.
+No big deal if it’s a small application and one person wrote all the code, but with even a
+few thousand lines of code or two developers, it’s very easy to miss that. And then this
+happens:
+
+```csharp
+// in one part of the codebase
+object GetMeAValue(sring url) {
+  if (IsMalformed(url)) {
+    throw new Exception($"The url `{url}` is malformed!");
+  }
+  
+  // do something else to load data from the URL
+  return data;
+}
+
+string RenderHtml(object toRender) {
+  // if toRender can't generate valid HTML, throw Error("invalid HTML");
+  // if it can, theRenderedHTML;
+}
+
+void WriteOutput(string html)
+{
+  // I/O
+}
+
+// somewhere else in the codebase -- throws an exception
+var badUrl = "http:/www.google.com";  // missing a slash
+var response = GetMeAValue(badUrl);  // throws here
+
+// we never get here, but it could throw too
+var htmlForPage = RenderHtml(value);
+
+// so we definitely can't get here safely
+WriteOutput(htmlForPage);
+```
+Notice: there’s no way for the caller to know that the function will throw. Perhaps you’re
+very disciplined and write good docstrings for every function – and moreover, perhaps
+everyone’s editor shows it to them and they pay attention to that briefly-available
+popover. More likely, though, this exception throws at runtime and probably as a result of
+user-entered data – and then you’re chasing down the problem through error logs. More, if
+you do want to account for the reality that any function anywhere in C♯ might
+actually throw, you’re going to write something like this:
+```csharp
+try
+{
+  var badUrl = "http:/www.google.com";  // missing a slash
+  var response = GetMeAValue(badUrl);  // throws here
+
+  // we never get here, but it could throw too
+  var htmlForPage = RenderHtml(value);
+
+  // so we definitely can't get here safely
+  WriteOutput(htmlForPage); 
+}
+catch(Exception exn)
+{
+  HandleErr(exn);
+}
+```
+This is like the first example but even worse for repetition!  Also, C♯ can’t help you
+here! There's no type signatures to say “This throws an exception!”
+
+Instead, we can use a `Result` to get us a container type, much like `Maybe`, to let us
+deal with this scenario. A `Result` is either an Ok wrapping around a value (like Just does)
```

Comment:
```suggestion
deal with this scenario. A `Result` is either an `Ok` wrapping around a value (like `Just` does)
```

Reply:
I intentionally didn't format these as code throughout the docs because there actually aren't any `Ok`,`Err`,`Just`, or `None` types.

Reply:
Ahhhhh, right. That's an interesting difference. 👍